In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


3.1 Implementation from Scratch Step - by - Step Guide:

3.1.1 Step -1- Data Understanding, Analysis and Preparations:

In this step we will read the data, understand the data, perform some basic data cleaning, and store everything
in the matrix as shown below.

• Requirements:

Dataset → student.csv

• Decision Process:

In this step we will define the objective of the task.

– Objective of the Task -
To Predict the marks obtained in writing based on the marks of Math and Reading.
• To - Do - 1:
1. Read and Observe the Dataset.
2. Print top(5) and bottom(5) of the dataset {Hint: pd.head and pd.tail}.
3. Print the Information of Datasets. {Hint: pd.info}.
4. Gather the Descriptive info about the Dataset. {Hint: pd.describe}
5. Split your data into Feature (X) and Label (Y).

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df= pd.read_csv('/content/drive/MyDrive/AI/Resources/Week5/student.csv')
df.head(5)

Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62

In [6]:
df.tail(5)

Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Math     1000 non-null   int64
 1   Reading  1000 non-null   int64
 2   Writing  1000 non-null   int64
dtypes: int64(3)
memory usage: 23.6 KB


In [8]:
df.describe()

Math      Reading      Writing
count  1000.000000  1000.000000  1000.000000
mean     67.290000    69.872000    68.616000
std      15.085008    14.657027    15.241287
min      13.000000    19.000000    14.000000
25%      58.000000    60.750000    58.000000
50%      68.000000    70.000000    69.500000
75%      78.000000    81.000000    79.000000
max     100.000000   100.000000   100.000000

In [23]:
X = df[['Math', 'Reading']]  # Independent variables
Y = df['Writing']           # Dependent variable

print("\nFeatures (X):")
print(X.head())

print("\nLabels (Y):")
print(Y.head())


Features (X):
   Math  Reading
0    48       68
1    62       81
2    79       80
3    76       83
4    59       64

Labels (Y):
0    63
1    72
2    78
3    79
4    62
Name: Writing, dtype: int64


• To - Do - 2:
1. To make the task easier - let’s assume there is no bias or intercept.
2. Create the following matrices:

Y = WTXW =





w1
w2
.
.
.
wd





, where W ∈ R
d

X =





x1,1 x1,2 · · · x1,n
x2,1 x2,2 · · · x2,n
.
.
.
.
.
.
.
.
.
.
.
.
xd,1 xd,2 · · · xd,n





, where X ∈ R
d×n

Y =





y1
y2
.
.
.
yn





, where Y ∈ R
n

3. Note: The feature matrix described above does not include a column of 1s, as it assumes the
absence of a bias term in the model.

In [24]:
X = X.values.T  # Transpose to shape d x n
print("\nFeature Matrix X (d x n):")
print(X)

d = X.shape[0]  # Number of rows
W = np.random.rand(d, 1)  # Random initialization
print("\nWeight Matrix W (d x 1):")
print(W)

Y = Y.values.reshape(-1, 1)  # Reshape to n x 1
print("\nTarget Vector Y (n x 1):")
print(Y)


Feature Matrix X (d x n):
[[48 62 79 ... 89 83 66]
 [68 81 80 ... 87 82 66]]

Weight Matrix W (d x 1):
[[0.53269773]
 [0.60190505]]

Target Vector Y (n x 1):
[[ 63]
 [ 72]
 [ 78]
 [ 79]
 [ 62]
 [ 85]
 [ 83]
 [ 41]
 [ 80]
 [ 77]
 [ 64]
 [ 90]
 [ 45]
 [ 77]
 [ 70]
 [ 46]
 [ 76]
 [ 44]
 [ 85]
 [ 72]
 [ 53]
 [ 66]
 [ 75]
 [ 49]
 [ 84]
 [ 83]
 [ 68]
 [ 66]
 [ 77]
 [ 78]
 [ 74]
 [ 83]
 [ 72]
 [ 65]
 [ 46]
 [ 66]
 [ 50]
 [ 79]
 [ 68]
 [ 46]
 [ 86]
 [ 70]
 [ 61]
 [ 53]
 [ 72]
 [ 75]
 [ 50]
 [ 77]
 [100]
 [ 81]
 [100]
 [ 87]
 [ 78]
 [ 48]
 [ 50]
 [ 44]
 [ 48]
 [ 43]
 [ 67]
 [ 78]
 [ 58]
 [ 91]
 [ 92]
 [ 78]
 [ 42]
 [ 85]
 [ 73]
 [ 83]
 [ 61]
 [ 58]
 [ 60]
 [ 55]
 [ 48]
 [ 62]
 [ 68]
 [ 59]
 [ 62]
 [ 48]
 [ 74]
 [ 63]
 [ 80]
 [ 79]
 [ 73]
 [ 79]
 [ 45]
 [ 67]
 [ 89]
 [ 77]
 [ 81]
 [ 88]
 [ 53]
 [ 68]
 [ 79]
 [ 77]
 [ 63]
 [ 73]
 [ 60]
 [ 67]
 [100]
 [ 79]
 [ 26]
 [ 51]
 [ 80]
 [ 57]
 [ 41]
 [ 78]
 [ 68]
 [ 49]
 [ 76]
 [ 41]
 [ 71]
 [ 77]
 [ 89]
 [ 86]
 [ 55]
 [ 80]
 [ 56]
 [ 74]
 [ 85]
 [ 80]
 

In [25]:
Y_pred = np.dot(W.T, X)
print("\nPredicted Y (using initial W):")
print(Y_pred)


Predicted Y (using initial W):
[[ 66.49903445  81.78156834  90.23552476  90.4431467   69.95108933
   87.31616762  87.84886535  53.39553803  79.44315546  92.76059879
   74.55870777  91.32188101  56.31489517  90.09711013  74.62791508
   52.58601104  74.69712239  52.86284029  91.99299337  85.09520858
   70.06854317  70.08950396  79.02791158  64.85901968  87.98727998
   90.39490017  72.75299262  74.97395165  90.62980786  83.44886885
   84.16822775  90.9275979   78.63362847  82.01647603  57.17266869
   61.75300138  58.5358542   86.29901868  81.29711714  48.8571269
   91.25267369  87.89711188  71.94346563  56.66093174  73.42410498
   80.09330704  59.06855193  78.63362847 108.13330078  75.43744207
  113.46027811  92.45648379  81.78156834  54.66855544  55.13204586
   57.72632721  66.13203709  56.19744132  82.54917376  86.04315021
   65.06664163  99.05647853 112.25646802  88.31235577  58.90917652
  101.7199672   70.76061632 101.09710137  72.33774873  70.06854317
   72.93965378  66.94156409  56

• To - Do - 3:
1. Split the dataset into training and test sets.
2. You can use an 80-20 or 70-30 split, with 80% (or 70%) of the data used for training and the rest
for testing.

In [61]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

# Print shapes for verification
print("\nShapes after manual split (80-20):")
print(f"X_train: {X_train.shape}, Y_train: {Y_train.shape}")
print(f"X_test: {X_test.shape}, Y_test: {Y_test.shape}")



Shapes after manual split (80-20):
X_train: (80, 3), Y_train: (80,)
X_test: (20, 3), Y_test: (20,)


3.1.2 Step -2- Build a Cost Function:
Cost function is the average of loss function measured across the data point. As the cost function for Regression
problem we will be using Mean Square Error which is given by:

L(w) = 1
2n
Xn
i=1
ypred(i) − yi
2

where:
ypred(w) = WTX

In [62]:
def cost_function(X, Y, W):
    """
    Parameters:
    X: Feature Matrix (d x n, where d is number of features, n is number of samples)
    Y: Target Matrix (1 x n, where n is number of samples)
    W: Weight Matrix (d x 1, where d is number of features)

    Output:
    cost: Accumulated mean square error
    """
    n = len(Y) # no of clumns

    # Reshape W to (d, 1) to make sure it can be multiplied with X
    W = W.reshape(-1, 1)  # Shape: (2, 1)

    # Predicted values: h_theta = X * W
    y_pred = np.dot(X, W)  # Shape: (n, 1)

    # Calculate the squared error: (y_pred - Y)^2
    error = y_pred - Y.reshape(-1, 1)  # Reshape Y to (n, 1)
    squared_error = np.square(error)  # Element-wise square, Shape: (n, 1)

    # Mean Squared Error: (1 / (2n)) * sum(squared_error)
    cost = (1 / (2 * n)) * np.sum(squared_error)

    return cost

Designing a Test Case for Cost Function:

We will first calculate the loss value manually and then verify the output via our code. If the computed value
matches, we will proceed further.

In [63]:
# Test case
X_test = np.array([[1, 2], [3, 4], [5, 6]])
Y_test = np.array([3, 7, 11])
W_test = np.array([1, 1])
cost = cost_function(X_test, Y_test, W_test)
if cost == 0:
  print("Proceed Further")
else:
  print("something went wrong: Reimplement a cost function")
print("Cost function output:", cost_function(X_test, Y_test, W_test))

Proceed Further
Cost function output: 0.0


3.1.3 Step -3- Gradient Descent for Simple Linear Regression:

In [64]:
def gradient_descent(X, Y, W, alpha, iterations):
    """
    Perform gradient descent to optimize the parameters of a linear regression model.

    Parameters:
    X (numpy.ndarray): Feature matrix (m x d), where m is the number of samples, d is the number of features.
    Y (numpy.ndarray): Target vector (m x 1).
    W (numpy.ndarray): Initial guess for parameters (d x 1).
    alpha (float): Learning rate.
    iterations (int): Number of iterations for gradient descent.

    Returns:
    W_update (numpy.ndarray): Optimized parameters (d x 1).
    cost_history (list): List of cost values over iterations.
    """
    # Initialize cost history
    cost_history = [0] * iterations

    # Number of samples
    m = len(Y)

    for iteration in range(iterations):
        # Step 1: Hypothesis Values (Y_pred)
        Y_pred = np.dot(X, W)  # Shape: (m, 1)

        # Step 2: Compute the loss (difference between predicted and actual)
        loss = Y_pred - Y

        # Step 3: Compute the gradient (dw)
        dw = (1 / m) * np.dot(X.T, loss)  # Gradient (shape: d x 1)

        # Step 4: Update the weights
        W = W - alpha * dw  # Update rule

        # Step 5: Calculate the new cost and store it
        cost = cost_function(X, Y, W)  # Compute cost (MSE)
        cost_history[iteration] = cost  # Store cost for this iteration

    return W, cost_history

Test Code for Gradient Descent function:

In [56]:
# Generate random test data
np.random.seed(0) # For reproducibility
X = np.random.rand(100, 3) # 100 samples, 3 features
Y = np.random.rand(100)
W = np.random.rand(3) # Initial guess for parameters
# Set hyperparameters
alpha = 0.01
iterations = 1000
# Test the gradient_descent function
final_params, cost_history = gradient_descent(X, Y, W, alpha, iterations)
# Print the final parameters and cost history
print("Final Parameters:", final_params)
print("Cost History:", cost_history)

Final Parameters: [0.20551667 0.54295081 0.10388027]
Cost History: [0.10711197094660153, 0.10634880599939901, 0.10559826315680618, 0.10486012948320558, 0.1041341956428534, 0.10342025583900626, 0.1027181077540776, 0.1020275524908062, 0.10134839451441931, 0.1006804415957737, 0.1000235047554587, 0.09937739820884377, 0.09874193931205609, 0.09811694850887098, 0.09750224927850094, 0.0968976680842672, 0.09630303432313951, 0.09571818027612913, 0.09514294105952065, 0.09457715457692842, 0.09402066147216397, 0.09347330508290017, 0.09293493139511913, 0.09240538899833017, 0.09188452904154543, 0.0913722051899995, 0.09086827358260123, 0.09037259279010502, 0.08988502377398919, 0.08940542984603007, 0.08893367662855953, 0.08846963201539432, 0.08801316613342668, 0.08756415130486386, 0.08712246201010665, 0.08668797485125508, 0.08626056851623207, 0.08584012374351278, 0.08542652328745133, 0.08501965188419301, 0.0846193962181636, 0.08422564488912489, 0.08383828837978763, 0.08345721902397185, 0.08308233097530

3.1.4 Step -4- Evaluate the Model:

Evaluation in Machine Learning measures the goodness of fit of your build model. Lets see How Good is
model we designed above, as discussed in the class for regression we can use following function as evaluation
measure.
1. Root Mean Square Error:

In [58]:
def rmse(Y, Y_pred):
    """
    This function calculates the Root Mean Squared Error (RMSE).

    Parameters:
    Y (numpy.ndarray): Array of actual target values (n,).
    Y_pred (numpy.ndarray): Array of predicted values (n,).

    Returns:
    float: The RMSE value.
    """
    # Calculate the squared differences
    squared_errors = (Y - Y_pred) ** 2

    # Calculate the mean of squared errors
    mean_squared_error = np.mean(squared_errors)

    # Take the square root of the mean squared error
    rmse_value = np.sqrt(mean_squared_error)

    return rmse_value

2. R2 or Coefficient of Determination:

In [57]:
def r2(Y, Y_pred):
    """
    This function calculates the R-Squared (R²) value.

    Parameters:
    Y (numpy.ndarray): Array of actual target values (n,).
    Y_pred (numpy.ndarray): Array of predicted values (n,).

    Returns:
    float: R-Squared (R²) value.
    """
    # Step 1: Calculate the mean of actual values (ȳ)
    mean_y = np.mean(Y)

    # Step 2: Compute the Total Sum of Squares (SST)
    ss_tot = np.sum((Y - mean_y) ** 2)

    # Step 3: Compute the Sum of Squared Residuals (SSR)
    ss_res = np.sum((Y - Y_pred) ** 2)

    # Step 4: Calculate R-Squared (R²)
    r2 = 1 - (ss_res / ss_tot)

    return r2

3.1.5 Step -5- Main Function to Integrate All Steps:

In this section, we will create a main function that integrates the data loading, preprocessing, cost function,
gradient descent, and model evaluation. This will help in running the entire workflow with minimal effort.

• Objective:

The objective of the main function is to execute the full process, from loading the data to performing
linear regression using gradient descent and evaluating the results using metrics like RMSE and R2
.

• To - Do:
We will define a function that:
1. Loads the data and splits it into training and test sets.
2. Prepares the feature matrix (X) and target vector (Y).
3. Defines the weight matrix (W) and initializes the learning rate and number of iterations.
4. Calls the gradient descent function to learn the parameters.
5. Evaluates the model using RMSE and R2

In [60]:
from sklearn.model_selection import train_test_split
def main():
  # Step 1: Load the dataset
  data= pd.read_csv('/content/drive/MyDrive/AI/Resources/Week5/student.csv')
  # Step 2: Split the data into features (X) and target (Y)
  X = data[['Math', 'Reading']].values # Features: Math and Reading marks
  Y = data['Writing'].values # Target: Writing marks
  # Step 3: Split the data into training and test sets (80% train, 20% test)
  X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)
  # Step 4: Initialize weights (W) to zeros, learning rate and number of iterations
  W = np.zeros(X_train.shape[1]) # Initialize weights
  alpha = 0.00001 # Learning rate
  iterations = 1000 # Number of iterations for gradient descent
  # Step 5: Perform Gradient Descent
  W_optimal, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)
  # Step 6: Make predictions on the test set
  Y_pred = np.dot(X_test, W_optimal)
  # Step 7: Evaluate the model using RMSE and R-Squared
  model_rmse = rmse(Y_test, Y_pred)
  model_r2 = r2(Y_test, Y_pred)
  # Step 8: Output the results
  print("Final Weights:", W_optimal)
  print("Cost History (First 10 iterations):", cost_history[:10])
  print("RMSE on Test Set:", model_rmse)
  print("R-Squared on Test Set:", model_r2)

# Execute the main function
if __name__ == "__main__":
  main()

Final Weights: [0.34811659 0.64614558]
Cost History (First 10 iterations): [2013.165570783755, 1640.286832599692, 1337.0619994901588, 1090.4794892850578, 889.9583270083234, 726.8940993009545, 594.2897260808594, 486.4552052951635, 398.7634463599484, 327.4517147324688]
RMSE on Test Set: 5.2798239764188635
R-Squared on Test Set: 0.8886354462786421


The model's performance is acceptable. It has a high R-squared value (0.89), indicating that it explains most of the variance in the target variable, and the RMSE is reasonable (5.28). This suggests the model is neither overfitting nor underfitting, but rather providing good predictions on the test set.